### ISSUE TITLE
Address sanitizer error on SPHINCS+

### ISSUE THREAD

**Issue by claucece** [open-quantum-safe/liboqs#808]
On `SPHINCS+-Haraka-256f-robust`, `SPHINCS+-Haraka-256f-simple`, `SPHINCS+-Haraka-256s-robust`, `SPHINCS+-Haraka-256s-simple`,  and `SPHINCS+-Haraka-256f-robust`, some address sanitizer errors occur of the flavor:

```
. > ./build/tests/kat_sig SPHINCS+-Haraka-256s-robust
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.0/include/emmintrin.h:3553:10: runtime error: load of misaligned address 0x7ffee7648278 for type 'const __m128i' (vector of 2 'long long' values), which requires 16 byte alignment
0x7ffee7648278: note: pointer points here
 00 00 00 00  7c 99 35 a0 b0 76 94 aa  0c 6d 10 e4 db 6b 1a dd  2f d8 1a 25 cc b1 48 03  2d cd 73 99
              ^ 
UndefinedBehaviorSanitizer:DEADLYSIGNAL
==64875==ERROR: UndefinedBehaviorSanitizer: SEGV on unknown address 0x000000000000 (pc 0x00010964563a bp 0x7ffee76400d0 sp 0x7ffee763fc70 T265528)
==64875==The signal is caused by a READ memory access.
==64875==Hint: address points to the zero page.
    #0 0x109645639 in haraka_S_absorb (kat_sig:x86_64+0x10108e639)
    #1 0x109637e65 in PQCLEAN_SPHINCSHARAKA256SROBUST_AESNI_haraka_S (kat_sig:x86_64+0x101080e65)
    #2 0x109634acd in PQCLEAN_SPHINCSHARAKA256SROBUST_AESNI_tweak_constants (kat_sig:x86_64+0x10107dacd)
    #3 0x10972f239 in PQCLEAN_SPHINCSHARAKA256SROBUST_AESNI_initialize_hash_function (kat_sig:x86_64+0x101178239)
    #4 0x1097304b5 in PQCLEAN_SPHINCSHARAKA256SROBUST_AESNI_crypto_sign_signature (kat_sig:x86_64+0x1011794b5)
    #5 0x109616bf7 in OQS_SIG_sphincs_haraka_256s_robust_sign sig_sphincs_haraka_256s_robust.c:65
    #6 0x10a3c8c54 in OQS_SIG_sign sig.c:927
    #7 0x1085cb17b in sig_kat kat_sig.c:697
    #8 0x1085cb5df in main kat_sig.c:756
    #9 0x7fff660dd3d4 in start (libdyld.dylib:x86_64+0x163d4)

==64875==Register values:
rax = 0x00007ffee7648278  rbx = 0x0000000000000000  rcx = 0x00007fff66213be2  rdx = 0x0000000000000000  
rdi = 0x0000000000000000  rsi = 0x000000010bbffd00  rbp = 0x00007ffee76400d0  rsp = 0x00007ffee763fc70  
 r8 = 0x000000010b30f8b0   r9 = 0x0000000000000076  r10 = 0x00007fff66213be2  r11 = 0x0000000000000206  
r12 = 0x0000000000000000  r13 = 0x0000000000000000  r14 = 0x0000000000000000  r15 = 0x0000000000000000  
UndefinedBehaviorSanitizer can not provide additional info.
==64875==ABORTING
count = 0
seed = 061550234D158C5EC95595FE04EF7A25767F2E24CC2BC479D09D86DC9ABCFDE7056A8C266F9EF97ED08541DBD2E1FFA1
mlen = 33
msg = D81C4D8D734FCBFBEADE3D3F8A039FAA2A2C9957E835AD55B22E75BF57BB556AC8
pk = 3E784CCB7EBCDCFD45542B7F6AF778742E0F4479175084AA488B3B74340678AA66A812016C62357F407E92B7652E36E3F5CEB7E063236CA860F74B42B370DF45
sk = 7C9935A0B07694AA0C6D10E4DB6B1ADD2FD81A25CCB148032DCD739936737F2DB505D7CFAD1B497499323C8686325E4792F267AAFA3F87CA60D01CB54F29202A3E784CCB7EBCDCFD45542B7F6AF778742E0F4479175084AA488B3B74340678AA66A812016C62357F407E92B7652E36E3F5CEB7E063236CA860F74B42B370DF45
```

**Comment by dstebila** [comment]
@thomwiggers @joostrijneveld any thoughts?

**Comment by joostrijneveld** [comment]
Hmm, at first glance it seems like `haraka.c` is loading from input pointers that are not necessarily aligned (not just in haraka_S_absorb, but in many of the functions handling input pointers). A quick fix to verify whether that's the cause would be to change the LOAD macro (e.g., [as seen on this line](https://github.com/open-quantum-safe/liboqs/blob/master/src/sig/sphincs/pqclean_sphincs-haraka-256s-robust_aesni/haraka.c#L17)) to use `_mm_loadu_si128`, i.e.:

```
#define LOAD(src) _mm_loadu_si128((u128 *)(src))
```

Note that the `STORE` macro (one line below) does use the unaligned version of the `_mm_store` intrinsic.

Just blatantly switching to unaligned loads everywhere may come at a (small?) performance cost, but I'm not sure if there's a reasonable way to avoid that. Maybe @thomwiggers (or @kste ?) has further thoughts on that.

@claucece could you change the LOAD macro in the respective haraka.c files for variants relying on AESNI in your local copy, and see if you still run into the same errors?

**Comment by dstebila** [comment]
> Just blatantly switching to unaligned loads everywhere may come at a (small?) performance cost, but I'm not sure if there's a reasonable way to avoid that. Maybe @thomwiggers (or @kste ?) has further thoughts on that.

Intel's documentation does not report a performance difference for [_mm_loadu_si128](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_loadu_si128&expand=3416) versus [_mm_load_si128](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_load_si128&expand=3343).

**Comment by claucece** [comment]
> @claucece could you change the LOAD macro in the respective haraka.c files for variants relying on AESNI in your local copy, and see if you still run into the same errors?

I'll do this over the weekend, and investigate a little further to see if I find the root of the problem.

**Comment by claucece** [comment]
So, @joostrijneveld changing the macro to the unaligned load fix it everywhere. Not sure what performance impact it gives, though. I can try to look at other alternatives; but, if not, I can send the PR with the changes. Another things that happens is that some tests fail (with or without the change).. should I report them as well?

**Comment by joostrijneveld** [comment]
Sounds good! I was under the impression there would be a performance hit, but it appears that isn't the case anymore for recent CPUs.

Different failing tests sound like a separate issue; I'm sure reporting those will also be appreciated :+1:.

In any case, I'll port this to the SPHINCS+ repository as well. Thanks!

**Comment by dstebila** [comment]
Since we pull SPHINCS+ from PQClean, rather than merging the PR directly into liboqs, the preferred approach would be make the change in PQClean, and then liboqs will pull that from PQClean on our next sync.

**Comment by claucece** [comment]
Oh, right. I'll open it there then ;)

**Comment by joostrijneveld** [comment]
For reference: there's [a PR containing this change at the SPHINCS+ repository](https://github.com/sphincs/sphincsplus/pull/15)

**Comment by thomwiggers** [comment]
There's also a PR on PQClean now https://github.com/PQClean/PQClean/pull/315

**Comment by dstebila** [comment]
Thanks Thom. I'll pull it in once it lands.

**Comment by claucece** [comment]
Thank you, @thomwiggers , for that. I guess this can be closed now?

**Comment by dstebila** [comment]
Let's leave this open until #817 lands to make sure it resolves it.

**Comment by dstebila** [comment]
Resolved.
